File: sdks/python/src/stepflow_sdk/transport.py
# Licensed to the Apache Software Foundation (ASF) under one or more contributor
# license agreements.  See the NOTICE file distributed with this work for
# additional information regarding copyright ownership.  The ASF licenses this
# file to you under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License.  You may obtain a copy of
# the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations under
# the License.

from re import U
from typing import Any
from msgspec import Struct, Raw, UnsetType, UNSET

# Type alias for request IDs to match Rust RequestId enum
type RequestId = str | int


class RemoteError(Struct, kw_only=True):
    """
    The error that occurred during the method execution.
    """

    code: int
    """
    The error code.
    """

    message: str
    """
    The error message.
    """

    data: dict[str, Any] = {}
    """
    The error data.
    """


class Message(Struct, kw_only=True):
    """
    Message sent to request a method execution.
    """

    jsonrpc: str = "2.0"
    """
    The JSON-RPC version (must be "2.0")
    """

    id: str | UnsetType = UNSET
    """
    The request id. If not set, this is a notification.
    """

    method: str | UnsetType = UNSET
    """
    The method to execute.
    """

    params: Raw | UnsetType = UNSET
    """
    The parameters to pass to the method.
    """

    result: Raw | UnsetType = UNSET
    """
    The result of the method execution.
    """

    error: RemoteError | UnsetType = UNSET
    """
    The error that occurred during the method execution.
    """

File: stepflow-rs/crates/stepflow-isolation-proxy/src/vsock/transport.rs
// Copyright 2025 DataStax Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
// in compliance with the License. You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied. See the License for the specific language governing permissions and limitations under
// the License.

//! Transport abstraction over vsock and Unix sockets.
//!
//! Supports two connection modes:
//! - **Direct Unix socket** (dev/subprocess mode): plain `UnixStream::connect`
//! - **Firecracker vsock** (VM mode): connect to vsock UDS, send `CONNECT <port>\n`,
//!   read `OK <port>\n` response, then use the socket as a data pipe.

use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::net::UnixStream;

/// A connection to a worker over vsock or Unix socket.
pub struct VsockConnection {
    pub stream: UnixStream,
}

impl VsockConnection {
    /// Connect to a worker via a plain Unix domain socket (dev/subprocess mode).
    pub async fn connect(path: &str) -> std::io::Result<Self> {
        let stream = UnixStream::connect(path).await?;
        Ok(Self { stream })
    }

    /// Connect to a worker inside a Firecracker VM via the vsock UDS.
    ///
    /// Firecracker's host-initiated vsock protocol:
    /// 1. Connect to the vsock UDS path
    /// 2. Send `CONNECT <guest_port>\n`
    /// 3. Receive `OK <host_port>\n`
    /// 4. Connection is now a data pipe to the guest
    pub async fn connect_vsock(uds_path: &str, guest_port: u32) -> std::io::Result<Self> {
        let stream = UnixStream::connect(uds_path).await?;

        // We need to split into read/write halves for the handshake,
        // then reconstruct the stream for data transfer.
        let (reader, mut writer) = tokio::io::split(stream);

        // Send CONNECT handshake
        let connect_msg = format!("CONNECT {guest_port}\n");
        writer.write_all(connect_msg.as_bytes()).await?;
        writer.flush().await?;

        // Read response line
        let mut buf_reader = BufReader::new(reader);
        let mut response = String::new();
        buf_reader.read_line(&mut response).await?;

        if !response.starts_with("OK") {
            return Err(std::io::Error::new(
                std::io::ErrorKind::ConnectionRefused,
                format!(
                    "Firecracker vsock CONNECT {guest_port} failed: {}",
                    response.trim()
                ),
            ));
        }

        // Reunite the split halves back into a UnixStream
        let reader = buf_reader.into_inner();
        let stream = reader.unsplit(writer);

        Ok(Self { stream })
    }
}
